config/directory.go

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package config
+
+import (
+	"path/filepath"
+	"strconv"
+	"testing"
+)
+
+// TestStepConfigFunc is the callback type used with acceptance tests to
+// specify a string which identifies a directory containing Terraform
+// configuration files.
+type TestStepConfigFunc func(TestStepConfigRequest) string
+
+// TestStepConfigRequest defines the request supplied to types
+// implementing TestStepConfigFunc.
+type TestStepConfigRequest struct {
+	StepNumber int
+}
+
+// Exec executes TestStepConfigFunc if it is not nil, otherwise an
+// empty string is returned.
+func (f TestStepConfigFunc) Exec(req TestStepConfigRequest) string {
+	if f != nil {
+		return f(req)
+	}
+
+	return ""
+}
+
+// StaticDirectory is a helper function that returns the supplied
+// directory when TestStepConfigFunc is executed.
+func StaticDirectory(directory string) func(TestStepConfigRequest) string {
+	return func(_ TestStepConfigRequest) string {
+		return directory
+	}
+}
+
+// TestNameDirectory returns the name of the test when TestStepConfigFunc
+// is executed. This facilitates a convention of naming directories
+// containing Terraform configuration files with the name of the test.
+func TestNameDirectory(t *testing.T) func(TestStepConfigRequest) string { //nolint:paralleltest //Not a test
+	return func(_ TestStepConfigRequest) string {
+		return t.Name()
+	}
+}
```

Comment:
If we can thread `t.Name()` through to `TestStepConfigRequest` like I think we can, we should be able to drop the `testing.T` here. It'd also be awesome to show an example of what happens from the developer perspective in the Go documentation. 👍 

One 🌶️  take that I have is that I think we should encourage folks to use the Go conventional `testdata` directory as well, which I'm showing in the suggestion here. Reference: https://pkg.go.dev/cmd/go#hdr-Test_packages / https://github.com/golang/go/issues/14715 / https://go-review.googlesource.com/c/go/+/23353/3/src/cmd/go/test.go

```suggestion
// TestNameDirectory returns the name of the test when TestStepConfigFunc
// is executed. This facilitates a convention of naming directories
// containing Terraform configuration files with the name of the test.
//
// For example, given test code:
//
//    func TestExampleCloudThing_basic(t *testing.T) {
//        resource.Test(t, resource.TestCase{
//            Steps: []resource.TestStep{
//                {
//                    ConfigDirectory: config.TestNameDirectory(),
//                },
//            },
//        })
//    }
//
// The testing configurations will be expected in the
// testdata/TestExampleCloudThing_basic/ directory.
func TestNameDirectory() func(TestStepConfigRequest) string {
	return func(req TestStepConfigRequest) string {
		return filepath.Join("testdata", req.TestName)
	}
}
```